### PR TITLE
Admin sendgrid stats

### PR DESCRIPTION
**Pull request by Darshan-Chauhan**
Hello Team,

This PR adds Email Stats tab in Admin panel. Which shows statastics of different event related to email. 
Ticket link: https://trello.com/c/fE0sr3oo
Please review and let me know if any changes are required.

Thanks

### CODE REVIEW

File: app/assets/javascripts/active_admin.js [hunk 1]
```diff
@@ -7,8 +7,46 @@
...
+    }
+    if(!((date_regex.test(start)) && date_regex.test(end)))
+    {
+      message.innerHTML = "";
+      message.innerHTML = "Enter date in yyyy-mm-dd format.";
+      return false;
+    }
+    $.ajax({
+      data: {"start_date" : $('#start_date').val(),
+              "end_date" : $('#end_date').val()},
```
Comment (sampatbadhe) [inline app/assets/javascripts/active_admin.js]: @Darshan-Chauhan 

Can we make use of `start` and `end` variable which is defined above. 
also can change the name of those variable to  `start_date` and `end_date`

cc: @rohanpujaris

Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: Sure, will change it. @sampatbadhe

File: db/schema.rb [hunk 2]
```diff
@@ -13,128 +13,128 @@
...
-  create_table "active_admin_comments", force: true do |t|
-    t.string   "resource_id",   null: false
-    t.string   "resource_type", null: false
-    t.integer  "author_id"
-    t.string   "author_type"
-    t.text     "body"
+  create_table "active_admin_comments", force: :cascade do |t|
+    t.string   "resource_id",   limit: 255,   null: false
+    t.string   "resource_type", limit: 255,   null: false
+    t.integer  "author_id",     limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan why we are limit this?
it means we wont have more than 9999 authors?
I don't think we need to enforce all such limits here

Comment (rohandaxini) [inline db/schema.rb]: even if we won't have 9999 authors - but we don't need to restrict ID columns with such things

File: db/schema.rb [hunk 3]
```diff
@@ -13,128 +13,128 @@
...
-    t.string   "encrypted_password",     default: "", null: false
-    t.string   "reset_password_token"
+  create_table "admin_users", force: :cascade do |t|
+    t.string   "email",                  limit: 255, default: "", null: false
+    t.string   "encrypted_password",     limit: 255, default: "", null: false
+    t.string   "reset_password_token",   limit: 255
     t.datetime "reset_password_sent_at"
     t.datetime "remember_created_at"
-    t.integer  "sign_in_count",          default: 0
+    t.integer  "sign_in_count",          limit: 4,   default: 0
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 4]
```diff
@@ -13,128 +13,128 @@
...
-    t.integer  "recipient_id"
-    t.string   "recipient_type"
+  create_table "activities", force: :cascade do |t|
+    t.integer  "trackable_id",     limit: 4
+    t.string   "trackable_type",   limit: 255
+    t.integer  "owner_id",         limit: 4
+    t.string   "owner_type",       limit: 255
+    t.string   "key",              limit: 255
+    t.text     "parameters",       limit: 65535
+    t.integer  "recipient_id",     limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 5]
```diff
@@ -13,128 +13,128 @@
...
-    t.integer  "trackable_id"
-    t.string   "trackable_type"
-    t.integer  "owner_id"
-    t.string   "owner_type"
-    t.string   "key"
-    t.text     "parameters"
-    t.integer  "recipient_id"
-    t.string   "recipient_type"
+  create_table "activities", force: :cascade do |t|
+    t.integer  "trackable_id",     limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 6]
```diff
@@ -13,128 +13,128 @@
...
+    t.string   "trackable_type",   limit: 255
+    t.integer  "owner_id",         limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 7]
```diff
@@ -13,128 +13,128 @@
...
     t.datetime "updated_at"
   end
 
   add_index "admin_users", ["email"], name: "index_admin_users_on_email", unique: true, using: :btree
   add_index "admin_users", ["reset_password_token"], name: "index_admin_users_on_reset_password_token", unique: true, using: :btree
 
-  create_table "attachments", force: true do |t|
-    t.integer  "context_id"
+  create_table "attachments", force: :cascade do |t|
+    t.integer  "context_id",              limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 8]
```diff
@@ -13,128 +13,128 @@
...
-    t.string   "attachable_file_name"
-    t.string   "attachable_content_type"
-    t.integer  "attachable_file_size"
-    t.string   "context_type"
-    t.integer  "user_id"
+    t.string   "attachable_file_name",    limit: 255
+    t.string   "attachable_content_type", limit: 255
+    t.integer  "attachable_file_size",    limit: 4
+    t.string   "context_type",            limit: 255
+    t.integer  "user_id",                 limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 9]
```diff
@@ -13,128 +13,128 @@
...
   end
 
   add_index "attachments", ["context_id"], name: "index_attachments_on_context_id", using: :btree
   add_index "attachments", ["user_id"], name: "index_attachments_on_user_id", using: :btree
 
-  create_table "comments", force: true do |t|
-    t.integer  "user_id"
-    t.text     "comment"
+  create_table "comments", force: :cascade do |t|
+    t.integer  "user_id",    limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: db/schema.rb [hunk 10]
```diff
@@ -13,128 +13,128 @@
...
+    t.text     "comment",    limit: 65535
     t.datetime "created_at"
     t.datetime "updated_at"
-    t.integer  "post_id"
+    t.integer  "post_id",    limit: 4
```
Comment (rohandaxini) [inline db/schema.rb]: @Darshan-Chauhan increase this limit to be 7 digits.

File: lib/sendgrid_data.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        end
+      end
+      email_stats
+    end
+
+    def get_sendgrid_data(start_date, end_date)
+      begin
+        uri = URI.parse("#{SENDGRID_URL}?start_date=#{start_date}&end_date=#{end_date}")
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = true
```
Comment (rohandaxini) [inline lib/sendgrid_data.rb]: @shivanibhanwal @Darshan-Chauhan  do we need this SSL true ? 
We don't use SSL for knowbuddy site. 
Does this SSL `true` something to do with webhook and sendgrid api url ?

File: lib/sendgrid_data.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+      stats = get_sendgrid_data start_date, end_date
+      %w(requests delivered opens clicks bounces spam_reports).each do |v|
+        stats.each do |value|
+          email_stats["#{v}_counts"] += value["stats"][0]["metrics"][v]
+        end
+      end
+      email_stats
+    end
+
+    def get_sendgrid_data(start_date, end_date)
```
Comment (rohandaxini) [inline lib/sendgrid_data.rb]: @Darshan-Chauhan should we rename this method `get_sendgrid_data` to `get_sendgrid_email_activity_stats` ?

File: lib/sendgrid_data.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    def default_stats
+      { "requests_counts" => 0, "delivered_counts" => 0,
+        "opens_counts" => 0, "clicks_counts" => 0, "bounces_counts" => 0,
+        "spam_reports_counts" => 0 }
+    end
+
+    def get_stats(start_date, end_date)
+      email_stats = default_stats
+      stats = get_sendgrid_data start_date, end_date
+      %w(requests delivered opens clicks bounces spam_reports).each do |v|
```
Comment (rohandaxini) [inline lib/sendgrid_data.rb]: do we really want to track `spam_reports` @shivanibhanwal  @Darshan-Chauhan ?

**Comment by rohandaxini** [comment]
@Darshan-Chauhan need to fix column level limits at few places.

File: lib/sendgrid_data.rb (hunk 13 above)
Comment (Darshan-Chauhan) [inline lib/sendgrid_data.rb]: @rohandaxini  Spam reports are shown in Sendgrid Dashboard along with other events..
trello ref: https://trello.com/c/C7WSMWJg/44-sendgrid-api-integration-for-emails

I can remove it if not needed

File: lib/sendgrid_data.rb (hunk 12 above)
Comment (Darshan-Chauhan) [inline lib/sendgrid_data.rb]: @rohandaxini  yes sure. I'll change the name

File: lib/sendgrid_data.rb (hunk 13 above)
Comment (rohandaxini) [inline lib/sendgrid_data.rb]: @Darshan-Chauhan okay sure, u can keep it then

File: app/helpers/application_helper.rb [hunk 14]
```diff
@@ -4,6 +4,12 @@ def avatar_url(user, size)
     "http://gravatar.com/avatar/#{gravatar_id}.png?s=#{size}"
   end
 
+  def find_percentage(stat_count, request_count)
```
Comment (sampatbadhe) [inline app/helpers/application_helper.rb]: @Darshan-Chauhan 

Can you please add specs for this method?

cc: @rohanpujaris

File: app/assets/javascripts/active_admin.js [hunk 15]
```diff
@@ -7,8 +7,47 @@
...
+    }
+    if(!((date_regex.test(start_date)) && date_regex.test(end_date)))
+    {
+      message.innerHTML = "";
+      message.innerHTML = "Enter date in yyyy-mm-dd format.";
+      return false;
+    }
+    $.ajax({
+      data: { "start_date" : $('#start_date').val(),
+              "end_date" : $('#end_date').val() },
```
Comment (sampatbadhe) [inline app/assets/javascripts/active_admin.js]: Hello @Darshan-Chauhan 

Still you have not replaced `$('#start_date').val()` with `start_date` and `$('#end_date').val()` with `end_date`

screenshot from 2016-05-10 18 35 19

cc: @rohanpujaris @shivanibhanwal

File: app/assets/stylesheets/active_admin.css.scss [hunk 16]
```diff
@@ -39,6 +64,255 @@ $wmd-preview-border-color-grey: #cccccc;
...
+*::before, *::after {
+    box-sizing: border-box;
+}
+*::before, *::after {
+    box-sizing: border-box;
+}
+
+.submit-dates {
+    border-radius: 0;
+    background-color: none;
```
Comment (sampatbadhe) [inline app/assets/stylesheets/active_admin.css.scss]: @Darshan-Chauhan 

Either use 2 space indentation or 4 space indentation. not combination of both in same file.

cc: @rohanpujaris

File: config/routes.rb [hunk 17]
```diff
@@ -16,12 +16,12 @@
 
   get "/admin/activity_types/deactivate/:id",
     to: "admin/activity_types#deactivate"
-
+  get "admin/email_stats/index", to: "admin/email_stats#index"
+  post "/admin/email_stats/index", to: "admin/email_stats#index"
```
Comment (sampatbadhe) [inline config/routes.rb]: Hello @Darshan-Chauhan 

Do we need both `get` and `post` requests here?

cc: @rohanpujaris

File: app/assets/javascripts/active_admin.js (hunk 15 above)
Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: @sampatbadhe I am changing it.

File: app/assets/stylesheets/active_admin.css.scss (hunk 16 above)
Comment (Darshan-Chauhan) [inline app/assets/stylesheets/active_admin.css.scss]: @sampatbadhe  changing it to 2 space indentation.

File: app/helpers/application_helper.rb (hunk 14 above)
Comment (Darshan-Chauhan) [inline app/helpers/application_helper.rb]: @sampatbadhe  I'll add the specs.

File: lib/sendgrid_data.rb (hunk 11 above)
Comment (Darshan-Chauhan) [inline lib/sendgrid_data.rb]: @rohandaxini , It is not related to webhooks and neither to SENDGRID_URL = "https://api.sendgrid.com/v3/stats"?

Without this line => http.use_ssl = true
it gives error = > 400 The plain HTTP request was sent to HTTPS port

ref: http://www.jjoe64.com/2012/02/ruby-https-request.html
      https://github.com/gitlabhq/gitlabhq/issues/3549

File: config/routes.rb (hunk 17 above)
Comment (Darshan-Chauhan) [inline config/routes.rb]: Hi @sampatbadhe 

I am thinking of keeping both get and post requests.
get for user to direct enter url and on selecting Start date and End date(Filters) I am submitting the page to the same Index method of email_stats.

File: app/helpers/application_helper.rb (hunk 14 above)
Comment (rohandaxini) [inline app/helpers/application_helper.rb]: @Darshan-Chauhan does this method `find_percentage ` need to be in `application_helper` ? Can u re-check please?
Also, rename method name `find_percentage` to be more specific and readable. For example `find_email_stats_percentage`

File: app/helpers/application_helper.rb [hunk 18]
```diff
@@ -4,6 +4,13 @@ def avatar_url(user, size)
...
+    return "0.00%" if request_count == 0 || stat_count == 0 ||
+                      request_count.nil? || stat_count.nil?
```
Comment (swatijadhav) [inline app/helpers/application_helper.rb]: This condition can be simply written as

        if request_count.to_i.zero? || stat_count.to_i.zero?

Comment (Darshan-Chauhan) [inline app/helpers/application_helper.rb]: Yes @swatijadhav , I'll change it.

File: app/views/admin/email_stats/_email_stats.html.haml [hunk 19]
```diff
@@ -0,0 +1,52 @@
+%ul.stats-design
+  %li.set-requests
+    REQUESTS
+    .request-counts
+      - if @email_stats["requests_counts"].nil?
+        = 0
+      = @email_stats["requests_counts"]
```
Comment (swatijadhav) [inline app/views/admin/email_stats/_email_stats.html.haml]: No need of `.nil?` check here, we can simply add it as

    .request-counts= @email_stats["requests_counts"].to_i

Comment (Darshan-Chauhan) [inline app/views/admin/email_stats/_email_stats.html.haml]: @swatijadhav ,
Sure , that'll work. I'll change it.

File: app/helpers/application_helper.rb (hunk 14 above)
Comment (rohanpujaris) [inline app/helpers/application_helper.rb]: @Darshan-Chauhan, I would suggest we keep it find_precentage and keep it in application_helper. Also we should find some generalized name to find_percentage arguments and make this method generic to be used anywhere. 

cc: @rohandaxini

File: app/assets/javascripts/active_admin.js [hunk 20]
```diff
@@ -7,8 +7,47 @@
...
+#= require constants
 
 $(document).ready(function(){
+  addDatepicker();
+  $('.stats-design').submit(function(event){
+    event.preventDefault();
+    var start_date = $('#start_date').val();
+    var end_date = $('#end_date').val();
+    var validation_message = $('#validation-error');
+    var date_regex = regex_for_date;
```
Comment (varunlalan) [inline app/assets/javascripts/active_admin.js]: Why add `regex_to_date` to another variable?

File: app/assets/javascripts/active_admin.js [hunk 21]
```diff
@@ -7,8 +7,47 @@
...
+    if(!((date_regex.test(start_date)) && date_regex.test(end_date)))
+    {
+      validation_message.html("Enter date in yyyy-mm-dd format.");
+      return false;
+    }
+    $.ajax({
+      data: { "start_date" : start_date,
+              "end_date" : end_date },
+      url:$('.dates').attr('action'),
+      method:$('.dates').attr('method'),
```
Comment (varunlalan) [inline app/assets/javascripts/active_admin.js]: `$(".dates")` can be cached.

File: lib/sendgrid_data.rb [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = true
+        request = Net::HTTP::Get.new(uri.request_uri)
+        request.add_field("Authorization", "Bearer #{SENDGRID_KEY['api_key']}")
+        response = http.request(request)
+        JSON.parse(response.body)
+      rescue StandardError => e
+        puts "Error Message: #{e.message}"
+      end
+    end
```
Comment (swatijadhav) [inline lib/sendgrid_data.rb]: Can we break this method further as follows?

    def get_sendgrid_email_activity_stats(start_date, end_date)
      uri = URI.parse("#{SENDGRID_URL}?start_date=#{start_date}&end_date=#{end_date}")
      JSON.parse(get(uri).body)
    end

    def get(uri)
      begin
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true
        request = Net::HTTP::Get.new(uri.request_uri)
        request.add_field("Authorization", headers)
        http.request(request)
      rescue StandardError => e
        puts "Error Message: #{e.message}"
      end
    end

    def headers
      "Bearer #{SENDGRID_KEY['api_key']}"
    end

File: app/assets/stylesheets/active_admin.css.scss [hunk 23]
```diff
@@ -28,6 +28,31 @@ $table-content-color-black: #000000;
 $wmd-preview-background-white: #ffffff;
 $wmd-preview-border-color-grey: #cccccc;
 
+$datepicker-background-color: #6a7176;
+$datepicker-title-color: #fff;
+$submit-dates-border-color: #00ccff;
+$submit-dates-background-color: #00ccff;
```
Comment (varunlalan) [inline app/assets/stylesheets/active_admin.css.scss]: Is it good to repeat same values in different variables?

cc: @rohanpujaris @sampatbadhe

File: app/helpers/application_helper.rb (hunk 14 above)
Comment (Darshan-Chauhan) [inline app/helpers/application_helper.rb]: @rohandaxini , it was written in application helper so it can be accessible everywhere. so I think we should keep it in application helper and also keep the name of method. as @rohanpujaris  suggested we can give generalized name to arguments.

File: db/schema.rb [hunk 24]
```diff
@@ -11,130 +11,136 @@
...
-    t.string   "author_type"
-    t.text     "body"
+ActiveRecord::Schema.define(version: 20160510144627) do
+
+  create_table "active_admin_comments", force: :cascade do |t|
+    t.string   "resource_id",   limit: 255,   null: false
+    t.string   "resource_type", limit: 255,   null: false
+    t.integer  "author_id",     limit: 8
+    t.string   "author_type",   limit: 255
+    t.text     "body",          limit: 65535
```
Comment (varunlalan) [inline db/schema.rb]: @Darshan-Chauhan I think you need to revert these unnecessary changes.

File: lib/sendgrid_data.rb (hunk 22 above)
Comment (Darshan-Chauhan) [inline lib/sendgrid_data.rb]: @swatijadhav , Yeah sure.

File: spec/classes/sendgrid_data_spec.rb [hunk 25]
```diff
@@ -0,0 +1,39 @@
+require 'rails_helper'
+require 'sendgrid_data'
+require 'json'
+
+describe 'SendgridData' do
+  describe 'class methods' do
+    describe 'get stats' do
+      it 'should not be empty' do
```
Comment (varunlalan) [inline spec/classes/sendgrid_data_spec.rb]: Avoid using `should` keyword while describing tests.

File: app/assets/stylesheets/active_admin.css.scss (hunk 23 above)
Comment (Darshan-Chauhan) [inline app/assets/stylesheets/active_admin.css.scss]: @varunlalan , I'll change it.

File: db/schema.rb (hunk 24 above)
Comment (Darshan-Chauhan) [inline db/schema.rb]: @varunlalan , I am keeping those changes as it was discussed with @rohandaxini .
cc: @rohandaxini

Comment (varunlalan) [inline db/schema.rb]: Sure thanks.

File: app/assets/javascripts/active_admin.js (hunk 20 above)
Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: @varunlalan ,
 regex_for_date variable was added in constants.js file.
As of there are no other constants in that file, I'll remove that file and will replace regex_for_date with regular expression.

File: app/assets/javascripts/active_admin.js (hunk 21 above)
Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: @varunlalan 
Yes, you're right. Will change it.

File: app/assets/javascripts/constants.js [hunk 26]
```diff
@@ -0,0 +1 @@
+var regex_for_date = /^[0-9]{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])$/;
```
Comment (rohanpujaris) [inline app/assets/javascripts/constants.js]: DATE_REGEX =  /^[0-9]{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])$/;

File: app/assets/javascripts/active_admin.js [hunk 27]
```diff
@@ -7,8 +7,47 @@
...
+    {
+      validation_message.html("Dates can't be empty or End Date must be greater than Start Date.");
+      return false;
+    }
+    if(!((date_regex.test(start_date)) && date_regex.test(end_date)))
+    {
+      validation_message.html("Enter date in yyyy-mm-dd format.");
+      return false;
+    }
+    $.ajax({
```
Comment (rohanpujaris) [inline app/assets/javascripts/active_admin.js]: Use if, else if and else instead of return false

File: app/assets/stylesheets/active_admin.css.scss [hunk 28]
```diff
@@ -40,6 +65,198 @@ $wmd-preview-border-color-grey: #cccccc;
...
+  text-shadow: 0 1px 0 #000;
+}
+
+*::before, *::after {
+  box-sizing: border-box;
+}
+*::before, *::after {
+  box-sizing: border-box;
+}
+
```
Comment (rohanpujaris) [inline app/assets/stylesheets/active_admin.css.scss]: Same rule written twice. Also check do we need this rule at all. If yes, please check that this does not impact any other pages. * means select everything

File: app/assets/stylesheets/active_admin.css.scss [hunk 29]
```diff
@@ -40,6 +65,198 @@ $wmd-preview-border-color-grey: #cccccc;
...
+form input[type="submit"], form input[type="button"], form button {
+  border-radius: 0;
+  background-color: $submit-dates-background-color;
+  background-image: none;
+  border-color: none;
+}
+
+form input[type="submit"]:hover:not(.disabled), form input[type="button"]:hover:not(.disabled), form button:hover:not(.disabled) {
```
Comment (rohanpujaris) [inline app/assets/stylesheets/active_admin.css.scss]: Break line after comma

File: app/assets/stylesheets/active_admin.css.scss (hunk 28 above)
Comment (Darshan-Chauhan) [inline app/assets/stylesheets/active_admin.css.scss]: @rohanpujaris , will check it than change it.

File: app/assets/stylesheets/active_admin.css.scss (hunk 29 above)
Comment (Darshan-Chauhan) [inline app/assets/stylesheets/active_admin.css.scss]: Ok. will do.

File: spec/classes/sendgrid_data_spec.rb [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+        allow(SendgridData).to receive(:get_sendgrid_email_activity_stats).with("2016-05-03",
+          "2016-05-04").and_return(JSON.parse(
```
Comment (rohanpujaris) [inline spec/classes/sendgrid_data_spec.rb]: break line at proper place. We can break from JSON.parse . Also do not add spaces after line break to indent it to upper brackets

File: app/assets/javascripts/active_admin.js [hunk 31]
```diff
@@ -7,8 +7,51 @@
...
+#= require constants
 
 $(document).ready(function(){
+  addDatepicker();
+  $('.stats-design').submit(function(event){
+    event.preventDefault();
+    var start_date = $('#start_date').val();
+    var end_date = $('#end_date').val();
+    var validation_message = $('#validation-error');
+    var date_regex = DATE_REGEX;
```
Comment (sampatbadhe) [inline app/assets/javascripts/active_admin.js]: Hello @Darshan-Chauhan 

Any reason behind this assignment of  `DATE_REGEX` to `date_regex` variable?

You can use `DATE_REGEX` direct instead of assigning that into variable. 

cc: @rohanpujaris @rohandaxini @shivanibhanwal @Manisha-Darade

Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: @sampatbadhe , 
Yes,I'll remove variable date_regex.

**Comment by rohandaxini** [comment]
@sampatbadhe @rohanpujaris is this PR good to merge next week to deploy to prod ?

File: app/assets/javascripts/active_admin.js [hunk 32]
```diff
@@ -7,8 +7,52 @@
...
+    }
+    if(!((DATE_REGEX.test(start_date)) && DATE_REGEX.test(end_date)))
+    {
+      validation_message.html("Enter date in yyyy-mm-dd format.");
+      return false;
+    }
+    var email_stat_form = $('.dates');
+    var waiting_period = setTimeout(function() {
+    validation_message.html("It's taking longer time than usual. You can either wait or reload the page.");
+    }, WAITING_TIME);
```
Comment (rohanpujaris) [inline app/assets/javascripts/active_admin.js]: Correct indentation. Can we handle timeout in different way other then setTimeout. We can handle it at server, where we are using Net::HTTP. Also if we want to do it at client side, I think we might have timeout options in ajax. Check documentation http://api.jquery.com/jquery.ajax/.

Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: @rohanpujaris 
Yes we can use timeout option.. I tried it before, but was facing some issues so instead of that I used setTimeout function.
And I'll correct the indentation.

**Comment by rohanpujaris** [comment]
@rohandaxini, Will review code and let you know the status of PR

File: app/assets/javascripts/active_admin.js (hunk 32 above)
Comment (rohanpujaris) [inline app/assets/javascripts/active_admin.js]: @Darshan-Chauhan, As discussed we can use setTimeOut as disable_with option is not playing well with ajax timeout. Also we are not aborting request so we don't need Net::HTTP timeout.

**Comment by rohanpujaris** [comment]
:+1:

File: app/assets/javascripts/active_admin.js (hunk 32 above)
Comment (Darshan-Chauhan) [inline app/assets/javascripts/active_admin.js]: @rohanpujaris . Ok.

**Comment by rohanpujaris** [comment]
Move all css and js for this feature into separate file

**Comment by rohanpujaris** [comment]
Closing this PR as new branch created from it have all the code.

New PR Link: https://github.com/rohandaxini/knowbuddy/pull/76

cc: @sampatbadhe, @rohandaxini

**Comment by sampatbadhe** [comment]
Closing this PR as new branch created from it have all the code.
New PR Link:  PR #76: User post mail notifications by Darshan-Chauhan in rohandaxini/knowbuddy on GitHub
cc: @rohanpujaris, @rohandaxini
